File: re/extract.py
#!/usr/bin/env python3
"""GDS -> netlist extractor for the Jane Street ASIC puzzle.

Approach: the GDS keeps standard cells as named references and routing as
top-level polygons plus via cells. Signal pins of every sky130 cell are the
labeled li1 shapes inside the cell definition. So connectivity is pure
geometry:
  - collect all top-level metal shapes (li1..met5, drawing+pin datatypes)
  - collect via-cell instances (each contributes pads on two layers which are
    trivially connected through the via)
  - collect per-instance transformed pin shapes on li1
  - union-find everything that touches on the same layer
"""
import json
import math
import sys
from collections import Counter, defaultdict

import numpy as np
import gdstk
from shapely.geometry import Polygon as SPoly, Point
from shapely.strtree import STRtree

LI1 = 67
METALS = (67, 68, 69, 70, 71, 72)  # li1, met1..met5
DRAW = 20
PIN = 16
LABEL = 5

# cells with no signal function
PHYS_CELLS = {"tapvpwrvgnd_1", "decap_3", "fill_1", "fill_2", "fill_4", "fill_8"}
POWER_PINS = {"VPWR", "VGND", "VPB", "VNB"}


def short_type(name):
    return name.split("__")[-1] if "__" in name else name


class DSU:
    def __init__(self):
        self.p = {}

    def find(self, x):
        p = self.p
        while p.setdefault(x, x) != x:
            p[x] = p[p[x]]
            x = p[x]
        return x

    def union(self, a, b):
        ra, rb = self.find(a), self.find(b)
        if ra != rb:
            self.p[ra] = rb


def cell_polys(cell, layer, dts):
    """polygons of `cell` on (layer, dt in dts) as numpy point arrays."""
    out = [p.points for p in cell.polygons if p.layer == layer and p.datatype in dts]
    for path in cell.paths:
        for pp in path.to_polygons():
            if pp.layer == layer and pp.datatype in dts:
                out.append(pp.points)
    return out


POLY, MET1 = 66, 68
LICON, MCON = (66, 44), (67, 44)


def build_pinmap(cell):
    """Extract per-pin routable geometry from a stdcell definition.

    A signal pin's routable copper can appear on BOTH li1 (67/20) and met1
    (68/20), and a single pin's li1 may be several islands joined only through
    the poly gate (li1->licon->poly->licon->li1) or through met1 (li1->mcon->
    met1->mcon->li1). Model the intra-cell graph over li1/met1/poly nodes with
    mcon and poly-licon contacts, find the component carrying each pin label,
    and return that component's shapes keyed by layer. The router may land on
    either layer, so emitting both is what makes top-level connectivity exact.
    """
    li = cell_polys(cell, LI1, (DRAW, PIN))
    m1 = cell_polys(cell, MET1, (DRAW, PIN))
    po = cell_polys(cell, POLY, (DRAW,))
    li_g = [SPoly(p) for p in li]
    m1_g = [SPoly(p) for p in m1]
    po_g = [SPoly(p) for p in po]
    licons = [SPoly(p) for p in cell_polys(cell, LICON[0], (LICON[1],))]
    mcons = [SPoly(p) for p in cell_polys(cell, MCON[0], (MCON[1],))]

    dsu = DSU()
    li_t = STRtree(li_g) if li_g else None
    m1_t = STRtree(m1_g) if m1_g else None
    po_t = STRtree(po_g) if po_g else None

    def within_layer(geoms, tree, tag):
        for i, g in enumerate(geoms):
            for j in tree.query(g, predicate="intersects"):
                dsu.union((tag, i), (tag, int(j)))
    if li_t: within_layer(li_g, li_t, "li")
    if m1_t: within_layer(m1_g, m1_t, "m1")
    if po_t: within_layer(po_g, po_t, "po")

    # mcon: li1 <-> met1
    for c in mcons:
        a = [int(j) for j in li_t.query(c, predicate="intersects")] if li_t else []
        b = [int(j) for j in m1_t.query(c, predicate="intersects")] if m1_t else []
        nodes = [("li", j) for j in a] + [("m1", j) for j in b]
        for n in nodes[1:]:
            dsu.union(nodes[0], n)
    # licon on poly: li1 <-> poly (skip diffusion contacts)
    for c in licons:
        on_poly = [int(j) for j in po_t.query(c, predicate="intersects")
                   if po_g[int(j)].intersection(c).area > 0.4 * c.area] if po_t else []
        if not on_poly:
            continue
        a = [int(j) for j in li_t.query(c, predicate="intersects")] if li_t else []
        nodes = [("po", j) for j in on_poly] + [("li", j) for j in a]
        for n in nodes[1:]:
            dsu.union(nodes[0], n)

    # attach labels (li1 label points)
    labels = [(lb.text, Point(lb.origin)) for lb in cell.labels
              if lb.layer == LI1 and lb.texttype == LABEL]
    root_name = {}
    for text, pt in labels:
        hit = None
        if li_t:
            for j in li_t.query(pt.buffer(0.01), predicate="intersects"):
                if li_g[int(j)].distance(pt) < 0.02:
                    hit = ("li", int(j)); break
        if hit is None:
            continue
        r = dsu.find(hit)
        if r in root_name and root_name[r] != text:
            raise RuntimeError(f"{cell.name}: component holds {root_name[r]} and {text}")
        root_name[r] = text

    pins = defaultdict(lambda: {LI1: [], MET1: []})
    for i, g in enumerate(li_g):
        r = dsu.find(("li", i))
        if r in root_name:
            pins[root_name[r]][LI1].append(li[i])
    for i, g in enumerate(m1_g):
        r = dsu.find(("m1", i))
        if r in root_name:
            pins[root_name[r]][MET1].append(m1[i])
    return {k: v for k, v in pins.items()}


def ref_transform(ref):
    """Return (matrix2x2, offset) mapping cell coords to top coords."""
    rot = ref.rotation or 0.0
    k = round(rot / (math.pi / 2)) % 4
    c, s = [(1, 0), (0, 1), (-1, 0), (0, -1)][k]
    R = np.array([[c, -s], [s, c]], dtype=float)
    S = np.array([[1, 0], [0, -1]], dtype=float) if ref.x_reflection else np.eye(2)
    M = R @ S
    if ref.magnification not in (None, 1.0):
        M = M * ref.magnification
    return M, np.asarray(ref.origin, dtype=float)


def extract(gds_path, out_json):
    lib = gdstk.read_gds(gds_path)
    cells = {c.name: c for c in lib.cells}
    top = lib.top_level()[0]
    print(f"top cell: {top.name}")

    # --- per-celltype pin maps ---
    pinmaps = {}
    via_defs = {}
    for name, cell in cells.items():
        if name.startswith("VIA_"):
            via_defs[name] = {
                ly: cell_polys(cell, ly, (DRAW,)) for ly in METALS
            }
        elif name.startswith("sky130"):
            pinmaps[name] = build_pinmap(cell)

    # --- collect shapes per layer: (points, owner) ---
    layer_shapes = defaultdict(list)  # layer -> list of (SPoly, owner)

    # top-level routing (drawing + pin purpose)
    for p in top.polygons:
        if p.layer in METALS and p.datatype in (DRAW, PIN):
            layer_shapes[p.layer].append((SPoly(p.points), ("w", id(p))))
    for path in top.paths:
        for pp in path.to_polygons():
            if pp.layer in METALS and pp.datatype in (DRAW, PIN):
                layer_shapes[pp.layer].append((SPoly(pp.points), ("w", id(pp))))

    # instances
    insts = []
    n_via = 0
    for ref in top.references:
        cname = ref.cell.name
        M, o = ref_transform(ref)
        if cname.startswith("VIA_"):
            owner = ("v", n_via)
            n_via += 1
            for ly, polys in via_defs[cname].items():
                for pts in polys:
                    layer_shapes[ly].append((SPoly(pts @ M.T + o), owner))
        elif cname.startswith("sky130"):
            st = short_type(cname)
            iid = len(insts)
            insts.append({
                "id": iid, "type": st,
                "x": round(float(o[0]), 4), "y": round(float(o[1]), 4),
                "rot": round((ref.rotation or 0) / (math.pi / 2)) % 4,
                "refl": bool(ref.x_reflection),
                "pins": {},
            })
            for pin, bylayer in pinmaps[cname].items():
                if pin in POWER_PINS:
                    continue
                owner = ("p", iid, pin)
                for ly, polys in bylayer.items():
                    for pts in polys:
                        layer_shapes[ly].append((SPoly(pts @ M.T + o), owner))
        # INTERNAL_* cells: no geometry on real layers -> skip

    # --- union-find over touching shapes per layer ---
    dsu = DSU()
    all_owners = set()
    for ly, shapes in layer_shapes.items():
        geoms = [g for g, _ in shapes]
        tree = STRtree(geoms)
        for i, (g, owner) in enumerate(shapes):
            all_owners.add(owner)
            for j in tree.query(g, predicate="intersects"):
                dsu.union(owner, shapes[int(j)][1])
        print(f"layer {ly}: {len(shapes)} shapes")
    group_size = Counter(dsu.find(o) for o in all_owners)

    # --- assign nets ---
    netid = {}

    def net_of(owner):
        r = dsu.find(owner)
        if r not in netid:
            netid[r] = len(netid)
        return netid[r]

    for inst in insts:
        cname = "sky130_fd_sc_hd__" + inst["type"]
        for pin in pinmaps[cname]:
            if pin in POWER_PINS:
                continue
            owner = ("p", inst["id"], pin)
            if group_size[dsu.find(owner)] <= 1:
                # never touched anything: dangling pin
                inst["pins"][pin] = None
            else:
                inst["pins"][pin] = net_of(owner)

    # --- ports from top-level labels ---
    ports = {}
    for lb in top.labels:
        if lb.texttype != LABEL or lb.layer not in METALS:
            continue
        pt = Point(lb.origin)
        best = None
        for g, owner in layer_shapes[lb.layer]:
            if g.distance(pt) < 0.005:
                best = owner
                break
        if best is None:
            print(f"WARNING: no shape for label {lb.text}")
            continue
        if lb.text in ("VPWR", "VGND"):
            ports.setdefault(lb.text, net_of(best))
        else:
            ports[lb.text] = net_of(best)

    result = {
        "top": top.name,
        "insts": insts,
        "ports": ports,
        "n_nets": len(netid),
    }
    with open(out_json, "w") as f:
        json.dump(result, f)
    logic = [i for i in insts if i["type"] not in PHYS_CELLS]
    print(f"instances: {len(insts)} ({len(logic)} with signal pins)")
    print(f"nets: {len(netid)}  ports: {ports}")
    return result


if __name__ == "__main__":
    extract(sys.argv[1], sys.argv[2])

File: ocaml/src/extract.ml
(* GDS -> netlist extraction, a port of re/extract.py.

   The standard cells keep their names in the GDS but the routing is anonymous
   metal, so connectivity is pure geometry: every top-level metal shape, every
   via pad and every placed cell-pin shape is union-found with whatever it
   touches on the same layer. *)

let li1 = 67
let met1 = 68
let poly = 66
let metals = [ 67; 68; 69; 70; 71; 72 ] (* li1, met1 .. met5 *)
let draw = 20
let pin_dt = 16
let label_tt = 5
let licon_layer, licon_dt = (66, 44)
let mcon_layer, mcon_dt = (67, 44)

(* Distances are in nanometres, which the extractor requires the file to use
   as its database unit. *)
let db_unit_um = 0.001
let label_snap_nm = 10  (* a pin label may sit this far from its li1 shape (Python: 0.01 um) *)
let port_snap_nm = 5    (* a port label from its top-level shape (Python: 0.005 um) *)
let index_pitch_nm = 2000

let is_power pin = List.mem pin [ "VPWR"; "VGND"; "VPB"; "VNB" ]
let is_rail name = name = "VPWR" || name = "VGND"

(* "sky130_fd_sc_hd__nand2_2" -> "nand2_2", scanning for the separator left to
   right like Python's name.split("__")[-1] *)
let short_type name =
  let n = String.length name in
  let rec go i start =
    if i + 1 >= n then start
    else if name.[i] = '_' && name.[i + 1] = '_' then go (i + 2) (i + 2)
    else go (i + 1) start
  in
  let start = go 0 0 in
  String.sub name start (n - start)

(* shapes of a cell on (layer, datatype in dts); one rectangle list per shape *)
let cell_shapes (cell : Gds.cell) layer dts =
  let polys =
    List.filter_map
      (fun (p : Gds.Polygon.t) ->
        if p.layer = layer && List.mem p.datatype dts then Some (Geom.of_polygon p.points)
        else None)
      cell.polygons
  in
  let paths =
    List.filter_map
      (fun (p : Gds.Path.t) ->
        if p.layer = layer && List.mem p.datatype dts then Some [ Geom.of_path p ] else None)
      cell.paths
  in
  polys @ paths

let shapes_touch a b = List.exists (fun ra -> List.exists (Geom.intersects ra) b) a

(* Per-pin routable copper of a standard cell, on li1 and met1.

   A pin's li1 may be several islands joined only through the poly gate
   (li1 -> licon -> poly -> licon -> li1) or through met1 (li1 -> mcon -> met1
   -> mcon -> li1), so the intra-cell graph over li1/met1/poly shapes is
   union-found through mcon and poly-landing licon contacts, and the component
   carrying each pin label is the pin. *)
let build_pinmap (cell : Gds.cell) : (string * (Geom.rect list * Geom.rect list)) list =
  let li = Array.of_list (cell_shapes cell li1 [ draw; pin_dt ]) in
  let m1 = Array.of_list (cell_shapes cell met1 [ draw; pin_dt ]) in
  let po = Array.of_list (cell_shapes cell poly [ draw ]) in
  let licons = List.concat (cell_shapes cell licon_layer [ licon_dt ]) in
  let mcons = List.concat (cell_shapes cell mcon_layer [ mcon_dt ]) in
  let nli = Array.length li and nm1 = Array.length m1 and npo = Array.length po in
  let dsu = Geom.Dsu.create (nli + nm1 + npo) in
  let node_li i = i and node_m1 j = nli + j and node_po k = nli + nm1 + k in
  let within arr node =
    Array.iteri
      (fun i a ->
        for j = i + 1 to Array.length arr - 1 do
          if shapes_touch a arr.(j) then Geom.Dsu.union dsu (node i) (node j)
        done)
      arr
  in
  within li node_li;
  within m1 node_m1;
  within po node_po;
  let touching arr node c =
    let acc = ref [] in
    Array.iteri (fun i s -> if List.exists (Geom.intersects c) s then acc := node i :: !acc) arr;
    List.rev !acc
  in
  let union_all = function
    | [] -> ()
    | first :: rest -> List.iter (fun n -> Geom.Dsu.union dsu first n) rest
  in
  (* mcon: li1 <-> met1 *)
  List.iter (fun c -> union_all (touching li node_li c @ touching m1 node_m1 c)) mcons;
  (* licon landing on poly (more than 40% of its area): li1 <-> poly. Contacts
     on diffusion are skipped. *)
  List.iter
    (fun c ->
      let on_poly = ref [] in
      Array.iteri
        (fun k s ->
          let a = List.fold_left (fun acc r -> acc + Geom.inter_area r c) 0 s in
          if a * 10 > 4 * Geom.area c then on_poly := node_po k :: !on_poly)
        po;
      if !on_poly <> [] then union_all (List.rev !on_poly @ touching li node_li c))
    licons;
  (* Pin labels sit on li1 shapes. The nearest shape within the snap distance
     is taken; the Python takes the first spatial-index hit, which is the same
     shape whenever the candidates touch each other, as they do in every
     shipped cell. *)
  let root_name = Hashtbl.create 16 in
  let snap2 = label_snap_nm * label_snap_nm in
  List.iter
    (fun (t : Gds.Text.t) ->
      if t.layer = li1 && t.texttype = label_tt then begin
        let best = ref None in
        Array.iteri
          (fun i s ->
            let d = Geom.dist2_to_rects s t.origin in
            if d <= snap2 then
              match !best with Some (_, bd) when bd <= d -> () | _ -> best := Some (i, d))
          li;
        match !best with
        | None -> ()
        | Some (i, _) -> (
            let r = Geom.Dsu.find dsu (node_li i) in
            match Hashtbl.find_opt root_name r with
            | Some other when other <> t.text ->
                Util.fail "%s: one component holds pins %s and %s" cell.name other t.text
            | _ -> Hashtbl.replace root_name r t.text)
      end)
    cell.texts;
  (* collect, keeping pins in order of first appearance *)
  let order = ref [] and pins = Hashtbl.create 8 in
  let add name layer rects =
    if not (Hashtbl.mem pins name) then begin
      order := name :: !order;
      Hashtbl.replace pins name (ref [], ref [])
    end;
    let lir, m1r = Hashtbl.find pins name in
    if layer = li1 then lir := !lir @ rects else m1r := !m1r @ rects
  in
  let name_of node = Hashtbl.find_opt root_name (Geom.Dsu.find dsu node) in
  Array.iteri (fun i s -> Option.iter (fun name -> add name li1 s) (name_of (node_li i))) li;
  Array.iteri (fun j s -> Option.iter (fun name -> add name met1 s) (name_of (node_m1 j))) m1;
  List.rev_map
    (fun name ->
      let lir, m1r = Hashtbl.find pins name in
      (name, (!lir, !m1r)))
    !order

type 'pins inst = {
  id : int;
  typ : string;  (* short cell type, e.g. "nand2_2" *)
  x : int;       (* origin, database units *)
  y : int;
  rot : int;     (* quarter turns *)
  refl : bool;
  pins : 'pins;
}

type netlist = {
  top : string;
  insts : (string * int option) list inst list;  (* pin -> net, None when dangling *)
  ports : (string * int) list;
  n_nets : int;
}

let extract ?(log = ignore) (lib : Gds.library) : netlist =
  if Float.abs (lib.db_unit_um -. db_unit_um) > 1e-12 then
    Util.fail "database unit is %g um; this extractor assumes %g um (1 nm)" lib.db_unit_um
      db_unit_um;
  let top =
    match Gds.top_cells lib with
    | [ c ] -> c
    | [] -> failwith "no top cell"
    | cs -> Util.fail "several top cells: %s" (String.concat ", " (List.map (fun c -> c.Gds.name) cs))
  in
  log (Printf.sprintf "top cell: %s" top.name);
  let pinmaps = Hashtbl.create 64 and via_defs = Hashtbl.create 16 in
  List.iter
    (fun (c : Gds.cell) ->
      if String.starts_with ~prefix:"VIA_" c.name then
        Hashtbl.replace via_defs c.name
          (List.map (fun ly -> (ly, List.concat (cell_shapes c ly [ draw ]))) metals)
      else if String.starts_with ~prefix:"sky130" c.name then
        Hashtbl.replace pinmaps c.name (build_pinmap c))
    lib.cells;
  (* owners: one id per routing shape, via instance, or (instance, pin) *)
  let n_owner = ref 0 in
  let new_owner () =
    let o = !n_owner in
    incr n_owner;
    o
  in
  let layer_shapes = Hashtbl.create 8 in
  List.iter (fun ly -> Hashtbl.replace layer_shapes ly (ref [])) metals;
  let add ly r o =
    let l = Hashtbl.find layer_shapes ly in
    l := (r, o) :: !l
  in
  let routable ly dt = List.mem ly metals && (dt = draw || dt = pin_dt) in
  List.iter
    (fun (p : Gds.Polygon.t) ->
      if routable p.layer p.datatype then begin
        let o = new_owner () in
        List.iter (fun r -> add p.layer r o) (Geom.of_polygon p.points)
      end)
    top.polygons;
  List.iter
    (fun (p : Gds.Path.t) ->
      if routable p.layer p.datatype then add p.layer (Geom.of_path p) (new_owner ()))
    top.paths;
  let placed = ref [] and n_inst = ref 0 in
  List.iter
    (fun (r : Gds.Reference.t) ->
      if r.mag <> 1.0 then
        Util.fail "reference to %s has magnification %g; only 1 is supported" r.cell r.mag;
      let place = Geom.transform ~reflect:r.reflect ~turns:(Gds.quarter_turns r) ~origin:r.origin in
      if String.starts_with ~prefix:"VIA_" r.cell then begin
        let o = new_owner () in
        List.iter
          (fun (ly, rects) -> List.iter (fun rc -> add ly (place rc) o) rects)
          (Hashtbl.find via_defs r.cell)
      end
      else if String.starts_with ~prefix:"sky130" r.cell then begin
        let id = !n_inst in
        incr n_inst;
        let pins =
          List.filter_map
            (fun (pin, (lir, m1r)) ->
              if is_power pin then None
              else begin
                let o = new_owner () in
                List.iter (fun rc -> add li1 (place rc) o) lir;
                List.iter (fun rc -> add met1 (place rc) o) m1r;
                Some (pin, o)
              end)
            (Hashtbl.find pinmaps r.cell)
        in
        let x, y = r.origin in
        placed :=
          { id; typ = short_type r.cell; x; y; rot = Gds.quarter_turns r; refl = r.reflect; pins }
          :: !placed
      end
      (* INTERNAL_* cells carry no geometry on real layers *))
    top.references;
  (* union-find owners that touch on any layer *)
  let dsu = Geom.Dsu.create !n_owner in
  List.iter
    (fun ly ->
      let shapes = Array.of_list (List.rev !(Hashtbl.find layer_shapes ly)) in
      let owners = Array.map snd shapes in
      let idx = Geom.build ~pitch:index_pitch_nm (Array.to_list (Array.map fst shapes)) in
      Array.iteri
        (fun i (r, _) ->
          Geom.query idx r (fun j -> if j > i then Geom.Dsu.union dsu owners.(i) owners.(j)))
        shapes;
      log (Printf.sprintf "layer %d: %d shapes" ly (Array.length shapes)))
    metals;
  let group_size = Array.make !n_owner 0 in
  for o = 0 to !n_owner - 1 do
    let r = Geom.Dsu.find dsu o in
    group_size.(r) <- group_size.(r) + 1
  done;
  let netid = Hashtbl.create 1024 in
  let net_of o =
    let r = Geom.Dsu.find dsu o in
    match Hashtbl.find_opt netid r with
    | Some n -> n
    | None ->
        let n = Hashtbl.length netid in
        Hashtbl.replace netid r n;
        n
  in
  (* nets are numbered in instance order, which keeps the JSON byte-identical
     to the Python extractor's *)
  let insts =
    List.map
      (fun inst ->
        let pins =
          List.map
            (fun (pin, o) ->
              (* a pin that never touched anything is dangling *)
              if group_size.(Geom.Dsu.find dsu o) <= 1 then (pin, None) else (pin, Some (net_of o)))
            inst.pins
        in
        { inst with pins })
      (List.rev !placed)
  in
  (* ports from top-level labels: for the power rails the first label wins,
     for signals the last one does *)
  let ports = ref [] in
  let snap2 = port_snap_nm * port_snap_nm in
  List.iter
    (fun (t : Gds.Text.t) ->
      if t.texttype = label_tt && List.mem t.layer metals then begin
        let shapes = List.rev !(Hashtbl.find layer_shapes t.layer) in
        match List.find_opt (fun (r, _) -> Geom.dist2 r t.origin < snap2) shapes with
        | None -> log (Printf.sprintf "WARNING: no shape for label %s" t.text)
        | Some (_, o) ->
            let net = net_of o in
            if not (List.mem_assoc t.text !ports) then ports := !ports @ [ (t.text, net) ]
            else if not (is_rail t.text) then
              ports := List.map (fun (k, v) -> if k = t.text then (k, net) else (k, v)) !ports
      end)
    top.texts;
  let n_nets = Hashtbl.length netid in
  log
    (Printf.sprintf "instances: %d  nets: %d  ports: %d" (List.length insts) n_nets
       (List.length !ports));
  { top = top.name; insts; ports = !ports; n_nets }

(* --- JSON in the same schema as the Python extractor ---------------------- *)

let json_string s =
  let b = Buffer.create (String.length s + 2) in
  Buffer.add_char b '"';
  String.iter
    (fun c ->
      match c with
      | '"' -> Buffer.add_string b "\\\""
      | '\\' -> Buffer.add_string b "\\\\"
      | '\n' -> Buffer.add_string b "\\n"
      | '\t' -> Buffer.add_string b "\\t"
      | c when Char.code c < 0x20 -> Buffer.add_string b (Printf.sprintf "\\u%04x" (Char.code c))
      | c -> Buffer.add_char b c)
    s;
  Buffer.add_char b '"';
  Buffer.contents b

(* nanometres as micrometres, shortest decimal *)
let um_of_nm v =
  let s = Printf.sprintf "%d.%03d" (abs v / 1000) (abs v mod 1000) in
  let n = ref (String.length s) in
  while !n > 0 && s.[!n - 1] = '0' do decr n done;
  let s = String.sub s 0 !n in
  let s = if s.[String.length s - 1] = '.' then s ^ "0" else s in
  if v < 0 then "-" ^ s else s

let to_json (nl : netlist) =
  let b = Buffer.create (1 lsl 16) in
  let add = Buffer.add_string b in
  add (Printf.sprintf "{\"top\": %s, \"insts\": [" (json_string nl.top));
  List.iteri
    (fun i inst ->
      if i > 0 then add ", ";
      add
        (Printf.sprintf
           "{\"id\": %d, \"type\": %s, \"x\": %s, \"y\": %s, \"rot\": %d, \"refl\": %b, \"pins\": {"
           inst.id (json_string inst.typ) (um_of_nm inst.x) (um_of_nm inst.y) inst.rot inst.refl);
      List.iteri
        (fun j (pin, net) ->
          if j > 0 then add ", ";
          add
            (Printf.sprintf "%s: %s" (json_string pin)
               (match net with Some n -> string_of_int n | None -> "null")))
        inst.pins;
      add "}}")
    nl.insts;
  add "], \"ports\": {";
  List.iteri
    (fun i (k, v) ->
      if i > 0 then add ", ";
      add (Printf.sprintf "%s: %d" (json_string k) v))
    nl.ports;
  add (Printf.sprintf "}, \"n_nets\": %d}" nl.n_nets);
  Buffer.contents b
